statsmodels/sandbox/mice/mice.py

```diff
@@ -0,0 +1,1167 @@
+        cov = self.results[vname].cov_params().copy()
+        cov *= scale_pert
+        norm = _multivariate_normal(mean=self.results[vname].params,
+                                    cov=cov)
+        self.params[vname] = norm.rvs()
+
+    def perturb_params(self, vname):
+
+        if self.perturbation_method[vname] == "gaussian":
+            self.perturb_gaussian(vname)
+        elif self.perturbation_method[vname] == "boot":
+            self.perturb_bootstrap(vname)
+        else:
+            raise ValueError("unknown perturbation method")
+
+    def impute(self, vname):
+        # Wrap this in case we later add additional imputation
+        # methods.
+        self.impute_pmm(vname)
+
+    def update(self, vname):
+        """
+        Update a single variable.  This is a two-step process in which
+        first the parameters are perturbed, then the missing values
+        are re-imputed.
+        """
+
+        self.perturb_params(vname)
+        self.impute(vname)
+
+    def impute_pmm(self, vname):
+        """
+        Use predictive mean matching to impute missing values.
+
+        Notes
+        -----
+        The `perturb_params` method must be called first to define the
+        model.
+        """
+
+        k_pmm = self.k_pmm
+
+        endog_obs, exog_obs, exog_miss =\
+                   self.get_split_data(vname)
+
+        # Predict imputed variable for both missing and nonmissing
+        # observations
+        model = self.models[vname]
+        pendog_obs = model.predict(self.params[vname], exog_obs)
+        pendog_miss = model.predict(self.params[vname],
+                                         exog_miss)
+
+        # Jointly sort the observed and predicted endog values for the
+        # cases with observed values.
+        ii = np.argsort(pendog_obs)
+        endog_obs = endog_obs[ii]
+        pendog_obs = pendog_obs[ii]
+
+        # Find the closest match to the predicted endog values for
+        # cases with missing endog values.
+        ix = np.searchsorted(pendog_obs, pendog_miss)
+
+        # Get the indices for the closest k_pmm values on
+        # either side of the closest index.
+        ixm = ix[:, None] +  np.arange(-k_pmm, k_pmm)[None, :]
+
+        # Account for boundary effects
+        msk = np.nonzero((ixm < 0) | (ixm > len(endog_obs) - 1))
+        ixm = np.clip(ixm, 0, len(endog_obs) - 1)
+
+        # Get the distances
+        dx = pendog_miss[:, None] - pendog_obs[ixm]
+        dx = np.abs(dx)
+        dx[msk] = np.inf
+
+        # Closest positions in ix, row-wise.
+        dxi = np.argsort(dx, 1)[:, 0:k_pmm]
+
+        # Choose a column for each row.
+        ir = np.random.randint(0, k_pmm, len(pendog_miss))
+
+        # Unwind the indices
+        jj = np.arange(dxi.shape[0])
+        ix = dxi[[jj, ir]]
+        iz = ixm[[jj, ix]]
+
+        imputed_miss = np.array(endog_obs[iz])
+
+        self.store_changes(vname, imputed_miss)
```

Comment:
general design question ( I don't have an overview yet)
Can we avoid that many methods work with sideeffects, i.e. changing attributes instead of returning?
I feel more comfortable (safer) if only the main function (like a fit method) changes state.